api/client/commands.go

```diff
@@ -139,10 +140,38 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 		if _, err := os.Stat(root); err != nil {
 			return err
 		}
-		filename := path.Join(root, "Dockerfile")
+
+		absRoot, err := filepath.Abs(root)
+		if err != nil {
+			return err
+		}
+
+		var filename string
+		origDockerfile := *dockerfileName
+
+		if cmd.IsSet("f") || cmd.IsSet("file") {
+			if filename, err = filepath.Abs(*dockerfileName); err != nil {
+				return err
+			}
+		} else {
+			filename = path.Join(absRoot, *dockerfileName)
+		}
```

Comment:
Before I make a change I want to get agreement on what I think you're asking for:
- change the default Dockerfile to be "CONTEXT/Dockerfile"  (keeps backwards compat)
- this will not only be a code change but a help-text change, so the end user will see:
         -f, --file="CONTEXT-ROOT/Dockerfile"    Name of the Dockerfile
- then if the user wants to be explicit and specify -f but still use the default value then they would need to include the path to the Dockerfile (including build context) on the -f arg.
Right?
Seems reasonable to me.